crates/viewer/re_viewer_context/src/cache/video_stream_cache.rs

```diff
@@ -0,0 +1,587 @@
+
+                // Note that the conversion of this time value is already handled by `VideoDataDescription::timescale`:
+                // For sequence time we use a scale of 1, for nanoseconds time we use a scale of 1_000_000_000.
+                let decode_timestamp = re_video::Time(time.as_i64());
+
+                // Samples within a chunk are expected to be always in order.
+                debug_assert!(decode_timestamp > previous_max_presentation_timestamp);
+                previous_max_presentation_timestamp = decode_timestamp;
+
+                let is_sync_result = re_video::is_sample_start_of_gop(
+                    &values[byte_offset..(byte_offset + byte_length)],
+                    codec,
+                );
+                // TODO(andreas): Do something with the error?
+                // Detecting too few GOPs is better than too many since it merely means that the decoder has more work to do to get to a given frame.
+                let is_sync = is_sync_result.unwrap_or(false);
+
+                if is_sync {
+                    // Last GOP extends until here now
+                    if let Some(last_gop) = gops.back_mut() {
+                        last_gop.sample_range.end = sample_idx;
+                    }
+
+                    // New gop starts at this frame.
+                    gops.push_back(re_video::demux::GroupOfPictures {
+                        decode_start_time: decode_timestamp,
+                        sample_range: sample_idx..(sample_idx + 1),
+                    });
+                }
+
+                Some(re_video::SampleMetadata {
+                    is_sync,
+
+                    // TODO(#10090): No b-frames for now. Therefore sample_idx == frame_nr.
+                    frame_nr: sample_idx as u32,
+                    decode_timestamp,
+                    presentation_timestamp: decode_timestamp,
+
+                    // Filled out later for everything but the last frame.
+                    duration: None,
+
+                    // We're using offsets directly into the chunk data.
+                    buffer_index,
+                    byte_offset: byte_offset as u32,
+                    byte_length: byte_length as u32,
+                })
+            }),
+    );
+
+    // Any new samples actually added? Early out if not.
+    if sample_base_idx == samples.next_index() {
+        return Ok(());
+    }
+
+    // Fill out durations for all new samples plus the first existing sample for which we didn't know the duration yet.
+    // (We set the duration for the last sample to `None` since we don't know how long it will last.)
+    // (Note that we can't use tuple_windows here because it can't handle mutable references)
+    for sample in sample_base_idx.saturating_sub(1)..samples.next_index() - 1 {
+        samples[sample].duration = Some(
+            samples[sample + 1].presentation_timestamp - samples[sample].presentation_timestamp,
+        );
+    }
```

Comment:
This looks like it would crash if `sample_base_idx - 1` has been dropped from `samples`